syntax = "proto2";

package pc_specification;

enum DataTypeCategory {
    BOOL = 0;
    BYTE = 1;
    WORD = 2;
    DWORD = 3;
    LWORD = 4;
    CHAR = 5;
    INT = 6;
    SINT = 7;
    DINT = 8;
    LINT = 9;
    UINT = 10;
    USINT = 11;
    UDINT = 12;
    ULINT = 13;
    REAL = 14;
    LREAL = 15;
    TIME = 16;
    DATE = 17;
    TIME_OF_DAY = 18;
    DATE_AND_TIME = 19;
    DERIVED = 20;
    POU = 21;
    ARRAY = 22; //used for internal purposes only
    NOT_ASSIGNED = 23; // used for internal purposes only
}

enum FieldInterfaceType {
    VAR_INPUT = 0;
    VAR_OUTPUT = 1;
    VAR_IN_OUT = 2;
    VAR = 3;
    VAR_TEMP = 4;
    VAR_EXTERNAL = 5;
    VAR_GLOBAL = 6;
    VAR_ACCESS = 7;
    VAR_EXPLICIT_STORAGE = 8;
    NA = 9;
}

enum MemType {
    INPUT_MEM = 0;
    OUTPUT_MEM = 1;
    RAM_MEM = 2;
}

enum TaskType {
    INTERVAL = 0;
    INTERRUPT = 1;
}

enum PoUType {
    FC = 0;
    FB = 1;
    PROGRAM = 2;
}

enum LogLevels {
    LOG_NONE = 0;
    LOG_INFO = 1;
    LOG_ERROR = 2;
    LOG_VERBOSE = 3;
};

message FieldStorageSpec {
    optional MemType mem_type = 1;
    optional int64 byte_offset = 2;
    optional int32 bit_offset = 3;
    optional string full_storage_spec = 4;
}

message DataTypeField {
    required string field_name = 1;
    required string field_datatype_name = 2;
    optional double range_min = 3;
    optional double range_max = 4;
    optional string initial_value = 5;
    optional int32 dimension_1 = 6; // must only be specified if this field is an array
    optional int32 dimension_2 = 7; // // must only be specified if this field is an 2d-array
    optional FieldInterfaceType intf_type = 8 [default = NA];

    //May only be specified if parent datatype is a POU category datatype
    optional FieldStorageSpec field_storage_spec = 9; 
}

message DataTypeSpec {
    optional double range_min = 1;
    optional double range_max = 2;
    optional string initial_value = 3;

    // Must only be specified if this datatype is supposed to be a typedef of an ARRAY of [datatype_category]. 
    // i.e if datatype_category is INT and if datatype_spec with dimension_1 = 10 is specified,
    // then this datatype is a typedef of an integer array of size 10 (ARRAY[10] of INT)
    optional int32 dimension_1 = 4; // must only be specified if this field is typedeffing an array
    optional int32 dimension_2 = 5; // // must only be specified if this datatype is typedeffing a 2d-array

}

message DataType {
    required string name = 1;
    // Note that the datatype_category cannot be specified as an ARRAY or NOT_ASSIGNED 
    // This datatype will be treated as a typedef of an array if dimensions are specified in datatype_spec
    required DataTypeCategory datatype_category = 2 [default =NOT_ASSIGNED];

    // must be specified only if datatype category is POU or DERIVED
    repeated DataTypeField datatype_field = 3;

    // must be specified only if datatype category is POU 
    optional PoUType pou_type = 4; 

    // datatype_spec may only be specified if category is not POU or DERIVED or ARRAY or NOT_ASSIGNED
    optional DataTypeSpec datatype_spec = 5;
}

message SFBSpecification {
    required string sfb_name = 1;
    required int64 exec_time_ns = 2;
}

message SFCSpecification {
    required string sfc_name = 1;
    required int64 exec_time_ns = 2;
}

message InstructionSpecification {
    required string ins_name = 1;
    required int64 exec_time_ns = 2;
}

message IntervalTaskParams {
    required int64 interval_ms = 1;
}

message InterruptTaskParams {
    required string trigger_variable_field = 1;
}

message TaskSpecification {
    required string task_name = 1;
    required int32 priority = 2;
    required TaskType type = 3;
    optional IntervalTaskParams interval_task_params = 4;
    optional InterruptTaskParams interrupt_task_params = 5;
}

message ProgramVariableInitialization {
    required string pou_variable_field_name = 1;
    required string mapped_variable_field_name = 2; 
}

message ProgramSpecification {
    required string program_name = 1;
    required string pou_variable_type = 2;
    required string task_name = 3;
    repeated ProgramVariableInitialization initialization_maps = 4;
}

message ResourceSpecification {
    required string resource_name = 1;
    optional int64 input_mem_size_bytes = 2;
    optional int64 output_mem_size_bytes = 3;
    optional DataType resource_global_var = 4;
    repeated DataType pou_var = 5;
    repeated TaskSpecification tasks = 6;
    repeated ProgramSpecification programs = 7;
    
}
message MachineSpecification {
    required int32 num_cpus = 1;
    optional int64 ram_mem_size_bytes = 2;
    repeated InstructionSpecification ins_spec = 3;
    repeated SFCSpecification sfc_spec = 4;
    repeated SFBSpecification sfb_spec = 5;
    repeated ResourceSpecification resource_spec = 6;
}



message Specification {
    optional string config_name = 1;
    optional LogLevels log_level = 2;
    optional string log_file_path = 3;
    optional bool enable_kronos = 4;
    required int32 run_time_secs = 5;
    repeated DataType datatype_declaration = 6;
    optional DataType config_global_pou_var = 7;
    optional DataType config_access_pou_var =8;
    required MachineSpecification machine_spec = 9;
}
